Initial Proxy Implementation

sverch opened the pull request:
This is a working signature proxy that will add the proper headers to authenticate with AWS.  This allows for the use of more standard tools to interact with the API, because it hides Amazon's proprietary signing process.

See: https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html

This project is heavily indebted to/partially ported from this project: https://github.com/okigan/awscurl

SYU15 commented on src/aws_signature_builder/mod.rs:
```diff
@@ -0,0 +1,390 @@
+    let new_headers = task_4_build_auth_headers_for_the_request(
+        amzdate, payload_hash, algorithm, credential_scope, signed_headers, signature, access_key,
+        security_token);
+    return new_headers;
```
Is there a link you can refer to so that ppl can see what the shape of the headers should be/what each of the params are? I see you have a test but might be helpful if someone is trying to debug.

sverch replied:
Yeah, you calling this out makes me realize that even I don't fully understand what headers are supposed to be there and why....  I ported this from https://github.com/okigan/awscurl without touching anything since I knew that project worked for me.

[These are the docs](https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html) that I found that most closely match what that code does, so I suspect that's where it was originally copied from.

I wouldn't even be surprised if it's not even consistent across different products.  This protocol is bizarre and looks like it was designed by ten different teams.

I think I'll just paste this response into a comment actually.  Even though it's not the best, it's probably helpful to at least know that this was mainly copy-pasta.

sverch replied:
I added a helper function to set the headers, and an example of how to call these functions, so hopefully that makes it clearer what should be set (since they're set in one place).  Also mentioned that this code was copied.

SYU15 commented on src/aws_signature_builder/mod.rs:
```diff
@@ -0,0 +1,390 @@
+    };
+
+    // Step 4: Create the canonical headers and signed headers. Header names and value must be
+    // trimmed and lowercase, and sorted in ASCII order.  Note that there is a trailing \n.
```
Why do you have to add thee a trailing `\n`?

sverch replied:
Actually, this makes it even more obvious that it was directly copied from https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html.  I bet only Amazon knows the answer to that...

I should probably put a note in the README and link to that page from many of these comments to make it clear this is the situation.

sverch replied:
I'm not shocked that the implementation of a proprietary protocol like this was copy/pasted from a docs page rather than implemented from an actual spec.  It makes me question how feasible it would be to actually do this by following the Amazon docs.

SYU15 commented on src/main.rs:
```diff
@@ -0,0 +1,116 @@
+        let mut headers = ::std::collections::HashMap::new();
+        for (key, value) in req.headers().iter() {
+            headers.insert(String::from(key.as_str()), String::from(value.to_str().unwrap()));
+        }
```
I think it'd be helpful to explain what you are doing here and why.

sverch replied:
This is quite bad, and really comes from the fact that the function takes the headers as a hashmap from string to string (mainly because it was a python dict before).

I like that the library has no dependencies, but maybe no one really cares about that.

Hmm, also now that I'm reading [the HeaderMap docs](https://hyper.rs/hyper/master/hyper/header/index.html#headermap), I can see:

> It is a multimap structure, where each header name may have multiple associated header values. Given this, some of the APIs diverge from HashMap.

That might be a good reason to just take a HeaderMap.

SYU15 commented on src/aws_signature_builder/mod.rs:
```diff
@@ -0,0 +1,390 @@
+    return (amzdate, datestamp)
+}
+
+pub fn generate_aws_signature_headers(
```
This function is a beast!

sverch replied:
Haha, yeah.  This is a big insane pile of arguments.  It could definitely be improved.

I think this is related to [this comment above](https://github.com/sverch/aws-signature-proxy/pull/1#discussion_r337242110).  I could probably combine these arguments into various structs, like `AWSCredentials` and maybe even the full Request object along with it.

sverch replied:
Updated!  Now it only takes three arguments instead of 15.  :)

SYU15 commented on src/main.rs:
```diff
@@ -0,0 +1,116 @@
+        if new_headers.contains_key("Authorization") {
+            req.headers_mut().insert(AUTHORIZATION,
+                HeaderValue::from_str(&new_headers["Authorization"]).unwrap());
+        }
```
It might be nice to have a reference to the docs or a comment to understand when someone should/shouldn't expect these headers or exist? Unless you think that's unnecessary/unimportant.

sverch replied:
I think this is related to [this earlier thread about the headers](https://github.com/sverch/aws-signature-proxy/pull/1#discussion_r337236271).  Much of this code was copy/pasted from the AWS docs into awscurl, and then translated line by line into rust, so that understanding might not even be possible.  I'll say all that in a comment here as well.

SYU15 commented on README.md:
```diff
@@ -1,2 +1,37 @@
-Cargo template for generating CLIs
+# AWS Sigature Proxy
+
+A proxy whose entire job is to sign requests destined for the AWS API.
```
I know this is a first PR, but would be nice to have a couple more sentences (even copy/pasted from your blog) about why this would be desirable to use and what this is trying to achieve.

sverch replied:
Yeah, good point!  I can probably summarize something here.

First attempt:

> # Motivation
>
> The AWS API requires a [non-standard signing process](https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html) invented by Amazon, which makes it incompatible with many standard tools, as you can see from the need for projects like [`awscurl`](https://github.com/okigan/awscurl), also known as "curl but it works with AWS".
> 
> This project aims to fix that by hiding the non-standard signing process behind a proxy, so that any clients can use standard tools like `curl` (example below) and [OpenAPI](https://github.com/OpenAPITools/openapi-generator/) with standard clients.
> 
> That will hopefully allow us to take advantage of powerful open source tools that would otherwise be difficult to use with this API.  If you are bored and want to know why the proxy seemed like a good idea over other alternatives, feel free to read [this post](https://shaunverch.com/butter/open-source/2019/09/27/butter-days-6.html).
>
> # Usage
> ... etc. ...

sverch commented:
Thanks for the review!  I addressed your comments I think and it ended up much better (at least the public interface).  The internal private functions are still kind of messy, but I'm going to call that good enough for now.
